File: src/agent/registry.ts
import { config } from "../config.js";
import { notionTools } from "../tools/notion.js";
import { gmailTools } from "../tools/gmail.js";
import { calendarTools } from "../tools/calendar.js";
import { githubTools } from "../tools/github.js";
import { granolaTools } from "../tools/granola.js";
import type { ToolDef, ToolHandler } from "./tool-types.js";

const PROJECT_NOTION_TOOLS = new Set([
  "notion_search_projects",
  "notion_create_project",
  "notion_update_project",
]);

const TASK_NOTION_TOOLS = new Set([
  "notion_search_tasks",
  "notion_create_task",
  "notion_update_task",
]);

function activeNotionTools(): Record<string, ToolDef> {
  return Object.fromEntries(
    Object.entries(notionTools).filter(([name]) => {
      if (PROJECT_NOTION_TOOLS.has(name)) {
        return Boolean(config.NOTION_PROJECTS_DB_ID);
      }
      if (TASK_NOTION_TOOLS.has(name)) {
        return Boolean(config.NOTION_TASKS_DB_ID);
      }
      return true;
    }),
  ) as Record<string, ToolDef>;
}

const allTools: Record<string, ToolDef> = {
  ...activeNotionTools(),
  ...gmailTools,
  ...calendarTools,
  ...githubTools,
  ...granolaTools,
};

export const toolRegistry: Record<string, ToolHandler> = Object.fromEntries(
  Object.entries(allTools).map(([k, v]) => [k, v.handler]),
);

export const toolSpecs = Object.values(allTools).map((t) => t.spec);

export type { ToolDef, ToolHandler, ToolSpec } from "./tool-types.js";

File: src/agent/registry.py
from collections.abc import Awaitable, Callable
from typing import Any

from src.tools.calendar import calendar_tools
from src.tools.connections import connections_tools
from src.tools.github import github_tools
from src.tools.gmail import gmail_tools
from src.tools.google_connect import google_connect_tools
from src.tools.granola import granola_tools
from src.tools.memory import memory_tools
from src.tools.work import work_tools

ToolHandler = Callable[[Any], Awaitable[Any]]
ToolDef = dict[str, Any]


def _all_tools() -> dict[str, ToolDef]:
    tools: dict[str, ToolDef] = {
        **work_tools,
        **connections_tools,
        **google_connect_tools,
        **gmail_tools,
        **calendar_tools,
        **github_tools,
        **granola_tools,
        **memory_tools,
    }
    from src.config import get_settings

    if get_settings().slack_enabled:
        from src.tools.slack_reminders import slack_reminder_tools

        tools.update(slack_reminder_tools)
    return tools


all_tools = _all_tools()
tool_registry: dict[str, ToolHandler] = {name: tool["handler"] for name, tool in all_tools.items()}
tool_specs = [tool["spec"] for tool in all_tools.values()]
